src/NServiceBus.Persistence.DynamoDB/Outbox/WriteRequestBatchExecutionExtensions.cs

```diff
@@ -0,0 +1,178 @@
+#nullable enable
+
+namespace NServiceBus.Persistence.DynamoDB
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Amazon.DynamoDBv2;
+    using Amazon.DynamoDBv2.Model;
+    using NServiceBus.Logging;
+
+    // While this class looks like it has some generic helper flavour it was deliberately kept as close
+    // as possible to the requirements of the outbox persister and can only be used in that context.
+    // Otherwise some significant refactoring would be required.
+    static class WriteRequestBatchExecutionExtensions
+    {
+        public static async Task BatchWriteItemWithRetries(this IAmazonDynamoDB dynamoDbClient,
+            List<List<WriteRequest>> writeRequestBatches, OutboxPersistenceConfiguration configuration, ILog logger,
+            Func<TimeSpan, CancellationToken, Task>? delayOnFailure = default,
+            TimeSpan? retryDelay = default, CancellationToken cancellationToken = default)
+        {
+            delayOnFailure ??= DefaultDelay;
+
+            var operationCount = writeRequestBatches.Count;
+            var batchWriteTasks = new Task[operationCount];
+            for (var i = 0; i < operationCount; i++)
+            {
+                batchWriteTasks[i] = dynamoDbClient.WriteBatchWithRetries(writeRequestBatches[i], i + 1, operationCount,
+                    configuration, logger, delayOnFailure, retryDelay,
+                    cancellationToken);
+            }
+            await Task.WhenAll(batchWriteTasks).ConfigureAwait(false);
+        }
+
+        static async Task WriteBatchWithRetries(this IAmazonDynamoDB dynamoDbClient, List<WriteRequest> batch,
```

Comment:
would this read a bit nicer if we move the config stuff and logger into the ctor?

Reply:
There is no ctor. This is a static extension method.